Spyglass core library

paulangton opened the pull request:
Core library separated out from #8830. This PR should merge second, and will fail until #8920 merges. The spyglass core contains:

- `Spyglass`
- `ArtifactFetcher` implementation
- `Artifact` implementation
- `viewers` package used by viewer implementers
- Tests

This PR also contains `placeholder.go`, an empty import file used to prevent Bazel from complaining about unused deps. It will be removed when #8922 merges.

This PR does not contain Spyglass' `README.md` because the README contains information and instructions that are dependent on #8922, like config and the default viewers.

/area prow
/assign @fejta @BenTheElder @stevekuznetsov @krzyzacy

Edit: #8920 has merged, barring further review this PR is all set to merge.

paulangton commented:
/assign @fejta @BenTheElder @stevekuznetsov @krzyzacy

paulangton commented:
/assign @krzyzacy

krzyzacy commented on prow/spyglass/gcsartifact_fetcher.go:
```diff
@@ -0,0 +1,238 @@
+		}
+		resp, err := c.Get(req.String())
+		if err != nil {
+			logrus.WithError(err).Error("Error in GCS XML API GET request")
```
hummm, break here? What's the expect behavior if the request didn't get through?

paulangton replied:
added retries, broke into separate function

stevekuznetsov replied:
On the log side of things -- you want to add fields to logs like this to identify the request that was being made

krzyzacy commented on prow/spyglass/podlogartifact.go:
```diff
@@ -0,0 +1,152 @@
+// NewPodLogArtifact creates a new PodLogArtifact
+func NewPodLogArtifact(jobName string, buildID string, podName string, sizeLimit int64, ja *jobs.JobAgent) *PodLogArtifact {
+	if jobName == "" {
+		logrus.Error("Must specify non-empty jobName")
```
(if the name is empty it will still return a valid artifact object?)

paulangton replied:
fixed in latest

krzyzacy commented on prow/spyglass/spyglasstests/gcsartifact_fetcher_test.go:
```diff
@@ -0,0 +1,100 @@
+//			t.Errorf("Case %s produced more artifacts than expected. Expected: %s\nActual: %s.", tc.name, tc.expectedArtifacts, actualArtifacts)
+//		}
+//	}
+//}
```
nit: clean up and put it into an issue

paulangton replied:
#8923, added inline as well

krzyzacy commented on prow/spyglass/spyglasstests/podlogartifact_test.go:
```diff
@@ -0,0 +1,54 @@
+		{
+			name:     "\"Job\" Podlog readall",
+			artifact: spyglass.NewPodLogArtifact("job", "123", "", 500e6, fakeJa),
+			expected: []byte("clusterA"),
```
eh why this gives back `clusterA`? :-)

paulangton replied:
this is taken directly from the original test written for the podlog endpoint in `deck/jobs/job_test.go` test `TestGetLog`

fg123 commented:
/meow

stevekuznetsov commented on prow/spyglass/gcsartifact.go:
```diff
@@ -0,0 +1,147 @@
+	reader, err := a.handle.NewRangeReader(context.Background(), offset, -1)
+	defer reader.Close()
+	if err != nil {
+		logrus.WithField("artifact", a.JobPath()).WithError(err).Error("Error getting range reader to artifact")
```
Should be consistent on:
 - actually logging on all errors
 - using fields on all log calls


We use the following pattern in other places to ensure the second:

```go
func (a *GCSArtifact) ReadTail(n int64) ([]byte, error) {
    // whatever
    logrus.WithFields(fieldsFor(a)).WithError(err).Error("whatever")
}

func fieldsFor(artifact GCSArtifact) logrus.Fields {
    // whatever
}
```

stevekuznetsov replied:
Errors can be `.Error()` but in general you want to think about -- if there is some pathological behavior, it would be great to be able to debug the issue without redeploying or changing code, etc, just from logs.

paulangton replied:
cool pattern, definitely wasnt aware of this one :+1:

stevekuznetsov commented on prow/spyglass/gcsartifact_fetcher.go:
```diff
@@ -0,0 +1,256 @@
+
+// NewGCSJobSource creates a new GCSJobSource from a given bucket and jobPath
+func NewGCSJobSource(src string) *GCSJobSource {
+	linkPrefix := "gs://"
```
Can you isolate this logic to a `parseGcsURI` or something  so we can have concise unit tests for it?

fejta replied:
Steve don't you and I both have instances of this running around somewhere? Here's mine, feel free tomove out of testgrid

https://github.com/kubernetes/test-infra/blob/7c80a99067aebd61c99e16e8ee8b88a435390fd0/testgrid/util/gcs/gcs.go#L46-L84

paulangton replied:
oh thats excellent, ill definitely be using that

stevekuznetsov replied:
Yeah I think I linked my version last time, too :) It's somewhere under pod-utils/gcs

stevekuznetsov commented on prow/spyglass/gcsartifact_fetcher.go:
```diff
@@ -0,0 +1,256 @@
+}
+
+// isGCSSource recognizes whether a source string references a GCS bucket
+func isGCSSource(src string) bool {
```
I still feel like this is a bad design and should be reconsidered right now before any more work is done ...

fejta replied:
What is bad about this design?

stevekuznetsov replied:
You take structured data -- artifact type, names, fields, stringly type it, then parse everything back out later. The impl invents a new URL format with a leading `prowjob` to identify pod logs... we should just use types.

stevekuznetsov replied:
(Also I've written this review like ten times now and I'm a little tired of rewriting it on every PR)

BenTheElder replied:
We're using string keys, because the keys are strings... they come from a URL segment. Go does not have string enums. We're not trying to shove some format like JSON into the URL.

stevekuznetsov commented on prow/spyglass/gcsartifact_fetcher.go:
```diff
@@ -0,0 +1,256 @@
+}
+
+// GCSMarker holds the starting point for the next paginated GCS query
+type GCSMarker struct {
```
Isolate your XML client impl into a separate client package

BenTheElder replied:
having one method use the XML rest API hardly seems like a client...?

stevekuznetsov commented on prow/spyglass/podlogartifact.go:
```diff
@@ -0,0 +1,154 @@
+		return &PodLogArtifact{}
+	}
+	if buildID == "" {
+		logrus.Error("Must specify non-empty buildID")
```
I don't understand this API -- return an error from `NewPodLogArtifact` if you are going to validate

paulangton replied:
will be fixed in latest

stevekuznetsov commented on prow/spyglass/podlogartifact.go:
```diff
@@ -0,0 +1,154 @@
+		}
+		return u.String()
+	}
+	logrus.Error("Error getting link to pod log, empty job name or build ID")
```
You should be able to validate the last two cases in the constructor, right?

paulangton replied:
yes, fix incoming

stevekuznetsov commented on prow/spyglass/podlogartifact.go:
```diff
@@ -0,0 +1,154 @@
+	// logs, err = a.ja.GetJobLogByPodName(a.podName) TODO I'd like to support this eventually
+	logs, err = a.ja.GetJobLog(a.name, a.buildID)
+	if err != nil {
+		logrus.WithField("artifactName", a.name).WithError(err).Error("Error getting pod logs")
```
Wait so if we fail to read the logs, `Size()` returns no error and a zero size on the artifact? Why do we want that?

paulangton replied:
added an error to `Size()`

stevekuznetsov commented on prow/spyglass/spyglass.go:
```diff
@@ -0,0 +1,231 @@
+
+	// Special-casing the fetching of in-cluster artifacts that havent been found yet
+	neededArtifactNames := neededArtifacts(foundArtifactNames)
+	for _, ga := range neededArtifactNames {
```
What does `ga` stand for?

paulangton replied:
was supposed to stand for guaranteed artifact, fixed in latest

stevekuznetsov commented on prow/spyglass/spyglass.go:
```diff
@@ -0,0 +1,231 @@
+	neededArtifactNames := neededArtifacts(foundArtifactNames)
+	for _, ga := range neededArtifactNames {
+		switch ga {
+		case "build-log.txt":
```
This seems like a super tenuous workaround for the design and seems fragile. Can we do this more explicitly? If one of our most common use-cases is a special case, doesn't sound like the core design is handling it well.

paulangton replied:
This is meant to special case artifacts that can be found in-cluster and also are not found in GCS. At the moment, this means only the pod logs. Because Prow knows about these artifacts, I think it is safe to special case them.

BenTheElder replied:
this is a well known artifact that all jobs should produce. abstracting over the sources should simplify some viewer logic... just display build-log, regardless of if it came from the pod directly or not.

fejta commented:
/lint

fejta commented on prow/spyglass/artifactfetchers.go:
```diff
@@ -0,0 +1,52 @@
+)
+
+// JobSource gets information about a location storing the results of a single Prow job
+type JobSource interface {
```
It is not obvious to me why we need to export the `spyglass.JobSource` interface to other packages and it doesn't look like this is happening. So can this just be `type jobSource interface`? 

Please review all your types and all your methods to determine which of these really need to be exported.

Please only export those which need to be exported and make the remainder lower case.

Also this isn't java, why do we need an interface in the first place?

paulangton replied:
I used an interface because there are potentially other sources that artifacts could come from. I wanted to implement them with some consistent operations instead of accessing straight from the fields of a gcs-based job source struct

fejta commented on prow/spyglass/artifactfetchers.go:
```diff
@@ -0,0 +1,52 @@
+	// JobPath returns the common prefix (excluding the name of the bucket) of all artifacts produced
+	// by this job
+	JobPath() string
+	BucketName() string
```
For example, this seems an especially GCS-centric abstraction for an interface.

I also dislike the idea that BucketName should be treated separately from the rest of the URI anywhere aside from the immediate interaction with GCS. Working with unix's `/foo/bar` is substantially nicer than windows' `c:\bar`

BenTheElder replied:
I don't think bucket name should be exported at all.

BenTheElder replied:
It's not clear why this type is exported at all? Seems like we should have a local implementation of `Artifact` which has a reasonable public API, and other details should be private / opaque...

paulangton replied:
Removing `BucketName` method and unexporting `JobSource`

fejta commented on prow/spyglass/gcsartifact.go:
```diff
@@ -0,0 +1,147 @@
+// GCSArtifact represents some output of a prow job stored in GCS
+type GCSArtifact struct {
+	// The handle of the object in GCS
+	handle *storage.ObjectHandle
```
godoc allows a leading `The`

fejta commented on prow/spyglass/gcsartifact.go:
```diff
@@ -0,0 +1,147 @@
+}
+
+// CanonicalLink gets the GCS web address of the artifact
+func (a *GCSArtifact) CanonicalLink() string {
```
Why does everything need accessor methods?

paulangton replied:
In the artifact case, to ensure that viewers have access to some consistent data about any given artifact.

fejta replied:
Consider starting everything unexported an then exporting them only when necessary in subsequent PRs.

fejta replied:
Fair enough. In general I would recommend waiting until the PR that uses them to export (this simplifies review).

Otherwise I cannot distinguish between "this needs to be exported" from "I'm exporting by default because I'm used to another language." This seems important given that this seemed to start with everything exported

fejta commented on prow/spyglass/gcsartifact_fetcher.go:
```diff
@@ -0,0 +1,256 @@
+}
+
+// Artifacts lists all artifacts available for the given job source.
+// Uses the GCS XML API because it is ~2x faster than the golang GCS library
```
Are we sure this is sufficient justification for rolling our own client? What does 2x faster mean?

krzyzacy replied:
I think the data comes from one of the 5k-node scalability job, which lead us to switch to use this XML API cc @BenTheElder found

paulangton replied:
@fejta 2x faster meaning listing all artifacts for a given job (i.e. with a given prefix) with the XML API is 2x faster than the go `storage` client

BenTheElder replied:
The JSON APIs spit out pretty-printed data (including newlines, indent etc.), with a ton of fields we don't use. The XML API spits out only a handful of fields per entry and does the entire thing as a single line (the newlines alone are like 20k bytes per call). It's something like 1/4 of the bytes per page of 1000 results on average. For jobs with many artifacts it dropped the times from something like 5s to scan to more like 1-2s.

stevekuznetsov replied:
The direction I gave Paul on this is to talk to the `storage` team and see if they are OK with adding client-side options to not pretty print, etc. We should fix this at the source.

BenTheElder replied:
The _server_ pretty prints (curl the endpoint and see for yourself) AND the JSON API includes more fields. I don't think they're going to substantially change the JSON API for our use case in a reasonable time period. 

We can revisit this later. We're wrapping a single, simple, paginated REST / SOAP API call, from an API so stable the schema is nearly a decade old.

BenTheElder replied:
The `storage` client also has good reasons for using the JSON API. And we have an unusual use case in listing a large number of keys. Normally if you needed to track lots of keys you'd do this in a separate database tracking your blobs.

fejta replied:
> page load time is also 2x faster

I get that 2x faster means 2x faster. What is the _impact_? Saving 5m->10m is a lot different than saving 100ms->200ms.

The scalability jobs are pretty ridiculous. I'm not 100% convinced that saving 3s when viewing this job is worth rolling our own client.

stevekuznetsov replied:
@BenTheElder there is no way to ask the GCS server to not pretty-print? That is what I meant by editing the `storage` client

stevekuznetsov replied:
Can't we just provide `false` to this API and save us hundreds of lines of code to maintain?

https://cloud.google.com/storage/docs/json_api/v1/parameters#pp

BenTheElder replied:
Did some more digging on this for actual numbers, and yes you can send `prettyPrint=false` in the advanced params. The XML is still  _much_ smaller and faster though:

JSON, pretty print:
```console
$ curl -o $HOME/gcs-pretty.json 'https://www.googleapis.com/storage/v1/b/kubernetes-jenkins/o?prefix=logs%2Fci-kubernetes-e2e-gce-large-performance%2F121&maxResults=1000'
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 1256k  100 1256k    0     0  4888k      0 --:--:-- --:--:-- --:--:-- 4888k
```

JSON, don't pretty print:
```console
$ curl -o $HOME/gcs.json 'https://www.googleapis.com/storage/v1/b/kubernetes-jenkins/o?prefix=logs%2Fci-kubernetes-e2e-gce-large-performance%2F121&maxResults=1000&prettyPrint=false'
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 1157k  100 1157k    0     0  4865k      0 --:--:-- --:--:-- --:--:-- 4865k
```
XML:
```console
$ curl -o $HOME/gcs.xml 'https://kubernetes-jenkins.storage.googleapis.com/?prefix=logs%2Fci-kubernetes-e2e-gce-large-performance%2F121&max-keys=1000'
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  330k  100  330k    0     0   959k      0 --:--:-- --:--:-- --:--:--  959k
```
---
They're ~1.2MB per page (non pretty printed JSON) vs ~330k per page (XML). Exact bytes:
```console
$ du -s -B1 $HOME/gcs.json
1187840 /usr/local/google/home/bentheelder/gcs.json

$ du -s -B1 $HOME/gcs.xml
339968  /usr/local/google/home/bentheelder/gcs.xml
```


The XML response is ~1/4 (28.62%) the size of even the non-pretty-printed JSON, because the JSON simply includes a lot more data, none of which we use. Spyglass also became subjectively much more usable for jobs with many artifacts after the switch.

BenTheElder replied:
Some jobs easily have 17k artifacts (/shrug) for which we can expect roughly 20.2 MB of data vs  roughly 5.7 MB. It's definitely noticeable.

stevekuznetsov replied:
OK, so use the `fields` parameter as exposed by the JSON API to restrict what we are asking for https://cloud.google.com/storage/docs/json_api/v1/parameters#fields

stevekuznetsov replied:
Look, it's even smaller :)

```
curl -o $HOME/gcs.json 'https://www.googleapis.com/storage/v1/b/kubernetes-jenkins/o?prefix=logs%2Fci-kubernetes-e2e-gce-large-performance%2F121&maxResults=1000&prettyPrint=false&fields=items(id)'
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  155k  100  155k    0     0   103k      0  0:00:01  0:00:01 --:--:--  103k
```

BenTheElder replied:
Sure, but we'd have to roll our own client, and the XML API must remain stable due to S3 compatibility, whereas the JSON API may not, I haven't looked. This was a pretty quick change to improve the experience.

BenTheElder commented on prow/spyglass/artifactfetchers.go:
```diff
@@ -0,0 +1,52 @@
+	// job source and artifact name.
+	Artifact(src JobSource, name string, sizeLimit int64) viewers.Artifact
+	// CreateJobSource tries to create a usable job source from the provided source string for this fetcher
+	CreateJobSource(src string) (JobSource, error)
```
It seems like instead `ArtifactFetcher` should be produced from some opaque-ish job source, the details of which are not further interacted with, only the methods to obtain Artifacts...?

paulangton replied:
CreateJobSource probably shouldnt be exported, although the `ArtifactFetcher` construction does need the `client` details (auth, etc), so I dont think we can produce it automatically from a job source

paulangton commented:
Thank you everyone for the reviews on Spyglass so far, I believe fb37f6a should have addressed all outstanding comments. Please add more review or /lgtm if people are satisfied. #8922 is next (and last!), and should be a bit more straightforward to review.

fejta commented on prow/spyglass/artifactfetchers.go:
```diff
@@ -0,0 +1,51 @@
+// jobSource gets information about a location storing the results of a single Prow job
+type jobSource interface {
+	// CanonicalLink returns a link to the storage location this job's artifacts.
+	CanonicalLink() string
```
Does an internal interface need exported methods?

It is also not at all obvious to me why we need to define the interface here as opposed to letting the consumer define the interface. Interface I does not need to be defined when authoring package X. Package Y can just later define the interface and if X implements those methods all is good.

fejta commented on prow/spyglass/artifactfetchers.go:
```diff
@@ -0,0 +1,51 @@
+)
+
+// jobSource gets information about a location storing the results of a single Prow job
+type jobSource interface {
```
> I used an interface because there are potentially other sources that artifacts could come from

Wait until that potentiality arrives before creating the abstraction? Until then we don't need the extract complexity from excess abstraction.

fejta commented on prow/spyglass/gcsartifact.go:
```diff
@@ -0,0 +1,162 @@
+
+// ReadAt reads len(p) bytes from a file in GCS at offset off
+func (a *GCSArtifact) ReadAt(p []byte, off int64) (n int, err error) {
+	var ctx context.Context
```
Seems like context should be part of GCSArtifact, which can set this to `context.Background()` initially and allow user to override as appropriate. Then just pass `a.context` whever needed

fejta commented on prow/spyglass/gcsartifact.go:
```diff
@@ -0,0 +1,162 @@
+// gzipped returns whether the file is gzip-encoded in GCS
+func (a *GCSArtifact) gzipped() bool {
+	attrs, err := a.handle.Attrs(context.Background())
+	if err != nil {
```
IMO it would be better form for the return signature of this function to be `(bool, error)` so the caller can decide how to handle this error

fejta commented on prow/spyglass/gcsartifact_fetcher.go:
```diff
@@ -0,0 +1,256 @@
+// Uses the GCS XML API because it is ~2x faster than the golang GCS library
+// for large number of artifacts. It should also be S3 compatible according to
+// the GCS api docs.
+func (af *GCSArtifactFetcher) Artifacts(src jobSource) []string {
```
For example, this ought to be a `gcsJobSource`. Despite the fact that my future `erickJobSource` implements the `jobSource` interface it is not something that `GCSArtifactFetcher` can work with.

fejta commented on prow/spyglass/spyglass.go:
```diff
@@ -0,0 +1,234 @@
+
+var (
+	// These are artifacts that we guarantee will be created with every job.
+	// If they are found ina storage location, the stored ones will be used,
```
in a

fejta commented on prow/spyglass/spyglass.go:
```diff
@@ -0,0 +1,234 @@
+var (
+	// These are artifacts that we guarantee will be created with every job.
+	// If they are found ina storage location, the stored ones will be used,
+	// otherwise we will try to fetch them from the job runtimeIn the future
```
runtime. In

fejta commented on prow/spyglass/spyglass.go:
```diff
@@ -0,0 +1,234 @@
+	// If they are found ina storage location, the stored ones will be used,
+	// otherwise we will try to fetch them from the job runtimeIn the future
+	// this might contain job metadata, podspec, etc.
+	guaranteedArtifacts = []string{"build-log.txt"}
```
Not 100% sure that these artifacts are guaranteed to exist. We have jobs right now that do not use bootstrap and do not use the decorators, thus they do not have a build-log.txt

paulangton replied:
A. Shouldn't we be uploading logs for those jobs then?
B. We gracefully handle missing logs for situations like prowjob no longer exists or job got TTL'd

fejta commented on prow/spyglass/gcsartifact.go:
```diff
@@ -0,0 +1,162 @@
+func (a *GCSArtifact) ReadAt(p []byte, off int64) (n int, err error) {
+	var ctx context.Context
+	if a.gzipped() {
+		return 0, viewers.ErrUnsupportedOp
```
More precise errors would be better ("gzipped gcs objects do not support a read offsets" is a lot more clear than "unsupported operation")

paulangton commented:
All lint warnings are "Golint comments: should have a package comment, unless it's in another file for this package." The package comment is in another file for this package, thus this passes lint, no?

fejta commented:
/lint

fejta commented:
/lint

fejta commented:
> The package comment is in another file for this package, thus this passes lint, no?

Weird. Maybe this is because you're adding multiple files? The package comment looks right to me

fejta commented on prow/spyglass/gcsartifact.go:
```diff
@@ -0,0 +1,185 @@
+	defer reader.Close()
+	if err != nil {
+		logrus.WithFields(fieldsFor(a)).WithError(err).Error("Error getting range reader to artifact")
+		return []byte{}, err
```
`return nil, err`

fejta commented on prow/spyglass/gcsartifact.go:
```diff
@@ -0,0 +1,185 @@
+	offset := size - n
+	reader, err := a.handle.NewRangeReader(a.ctx, offset, -1)
+	defer reader.Close()
+	if err != nil {
```
Personally I think you should own your exceptions.

So this should be something like:
`err := fmt.Errorf("failed range read: %v", err)`

rather than just returning someone else's error


The key thing that the `ReadTail()` caller needs to distinguish that they cannot do own their own is whether the err came from:
```
a.gzipped()
a.Size()
a.handle.NewRangeReader()
ioutil.ReadAll()
```

So you ought to encode this info in the error you return

fejta commented on prow/spyglass/gcsartifact_fetcher.go:
```diff
@@ -0,0 +1,257 @@
+	bodies := [][]byte{}
+	marker := GCSMarker{}
+	c := http.Client{}
+	if scheme != "https" {
```
Consider:
```
const (
  http = "http"
  https = "https"
)
```

As these are used multiple places (2 in this file, although that's generally the border between should at 3 uses vs pointless at 1. However this will also make unit testing easier and the code less error prone)

fejta commented on prow/spyglass/gcsartifact_fetcher_test.go:
```diff
@@ -0,0 +1,103 @@
+		},
+	}
+
+	for _, tc := range testCases {
```
Personally I like running everything in parallel, which is both faster and doesn't require me to both with the name:
```
for _, tc := range cases {
  t.Run(tc.name, func(t *testing.T) {
    artifact := testAf.blah()
    if artifact.JobPath() != tc.expectedJobPath {
      t.Errorf("expected artifact with job path %s but got %s", tc.expectedJobPath, artifact.JobPath())  // name not necessary
    }
  })
}
```

Not necessary but worth considering. Also running things in parallel forces you to design code that doesn't have global state.

fejta commented on prow/spyglass/podlogartifact.go:
```diff
@@ -0,0 +1,167 @@
+// ReadTail reads the last n bytes of the pod log
+func (a *PodLogArtifact) ReadTail(n int64) ([]byte, error) {
+	logs, err := a.ja.GetJobLogTail(a.name, a.buildID, n)
+	if err != nil {
```
Seems like this should return an error rather than creating a `bytes.Reader` with an invalid `logs` value.

fejta commented on prow/spyglass/podlogartifact.go:
```diff
@@ -0,0 +1,167 @@
+	var err error
+	logs, err = a.ja.GetJobLog(a.name, a.buildID)
+	if err != nil {
+		logrus.WithField("artifactName", a.name).WithError(err).Error("Error getting pod logs")
```
Don't swallow the error please. I am generally unhappy with this pattern of conflating empty bytes with a failed response

fejta commented on prow/spyglass/podlogartifact_test.go:
```diff
@@ -0,0 +1,60 @@
+	"testing"
+)
+
+func TestPodLogReadAll(t *testing.T) {
```
Please unit test the other non-trivial functions: ReadTail, ReadAtMost, ReadAll, NewPodLogArtifact,

fejta commented on prow/spyglass/podlogartifact.go:
```diff
@@ -0,0 +1,167 @@
+}
+
+// UseContext is unsupported for pod logs
+func (a *PodLogArtifact) UseContext(ctx context.Context) error {
```
A more natural pattern would be for `NewPodLogArtifact()` to accept a `context.Context` rather than have a `UseContext()`

paulangton replied:
The reason for `UseContext` is that the behavior of the context needs to be specified by the viewers, since they are the ones making the read calls. The viewers dont construct artifacts themselves, only receive a list of `Artifact`.

paulangton replied:
Also, if we specify context in the `Read` methods, we lose conformance to `io.ReaderAt`. I discussed this in Steve's (?) reviews a bit more in depth.

fejta replied:
It isn't clear to me why `Artifact` needs a `UseContext` interface instead of just ensuring that whatever creates the artifacts creates it with a context.

Is there any chance that multiple viewers read the same artifact? If so then that won't work with UseContext anyway.

paulangton replied:
`Artifacts` are created by the `ArtifactFetcher`. If we create them with a context there and dont have `UseContext`, there's no way for the `viewers` to use the context in any meaningful way. There is a chance of two viewers reading the same artifact, I'd imagine it would actually be quite common. Would it work to have a `WithContext` method that returns a new `Artifact` instead of mutating the current one?

paulangton replied:
It's not essential, but it was part of an earlier discussion about inserting context so that viewers could cancel read requests

fejta commented on prow/spyglass/viewers/viewers.go:
```diff
@@ -0,0 +1,201 @@
+	var linesInContents int64
+	artifactSize, err := a.Size()
+	if err != nil {
+		return []string{}, err
```
I would expect `return nil, err` in these kinds of situations (and please own your own errors)

fejta commented on prow/spyglass/viewers/viewers.go:
```diff
@@ -0,0 +1,201 @@
+func UnregisterViewer(viewerName string) error {
+	_, ok := viewHandlerRegistry[viewerName]
+	if !ok {
+		return ErrInvalidViewName
```
Why do we need to return an error if it isn't registered? `delete(map, missingKey)` will just silently return in this scenario: https://blog.golang.org/go-maps-in-action

paulangton replied:
good point, removed error

fejta commented on prow/spyglass/viewers/viewers.go:
```diff
@@ -0,0 +1,201 @@
+		lastRead = int64(numBytesRead)
+		lastOffset = offset
+		if err != nil {
+			if err != io.EOF {
```
Please combine these two if statements (`if err != nil && err != io.EOF`)

fejta commented on prow/spyglass/spyglass.go:
```diff
@@ -0,0 +1,234 @@
+	for _, ep := range s.Eyepieces {
+		jobSource, err := ep.createJobSource(src)
+		if err != nil {
+			if err == ErrCannotParseSource {
```
please move this above the other if statement

````
if err == ErrCannotParseSource {
  continue
} else if err != nil {
  // foo
}
```

fejta commented on prow/spyglass/spyglass.go:
```diff
@@ -0,0 +1,234 @@
+func (s *Spyglass) Views(matchCache map[string][]string) []Lens {
+	lenses := []Lens{}
+	for viewer, matches := range matchCache {
+		if len(matches) != 0 {
```
Please minimize indentation:

```
for v, m := range mc {
  if len(m) == 0 {
    continue
  }
  title, err := ...
  // ...
}
```

fejta commented on prow/spyglass/spyglass.go:
```diff
@@ -0,0 +1,234 @@
+	Eyepieces []ArtifactFetcher
+
+	// Ja contains information about the current jobs in deck
+	Ja *jobs.JobAgent
```
Don't think this is a great name. Also consider making this an interface, which will simplify unit testing.

paulangton replied:
Agree, took this naming convention from `deck/jobs/jobs.go` (?), will change

fejta commented on prow/spyglass/podlogartifact.go:
```diff
@@ -0,0 +1,167 @@
+	buildID   string
+	podName   string
+	sizeLimit int64
+	ja        *jobs.JobAgent
```
```
type podLogJobAgent interface {
  GetJobLog
  GetJobLogTail
}
```
and now you can easily unit test `PodLogArtifact`

paulangton replied:
And the advantage to using this versus my current fake `JobAgent` is that this one gives me more control over `GetJobLog.*` implementations?

fejta replied:
Right. You need a way to inject errors into GetJobLog and GetJobLogTail. This is easier with a spy / fake implementation.

fejta commented on prow/spyglass/podlogartifact_test.go:
```diff
@@ -0,0 +1,60 @@
+			expected: []byte("clusterB"),
+		},
+	}
+	for _, tc := range testCases {
```
Your test cases should include handling error scenarios. Aka ensure that if a.Size() or a.fetchLogs() fail then this returns an error.  If you accept my other suggestion to switch to using an interface then you can do something like:

```
type fakePodLogJobAgent struct {
  data []bytes 
  orMaybeData map[string][]bytes
}

func (fplja *fakePodLogJobAgent, n) GetJobLogTail(name, build string) ([]byte, error) {
    if name + "/" + build != expectedName { return nil, fmt.Errorf("bad name: %v", name) }
    if build == "error" { return nil, errors.New("injected GetJobLogTail error") }
    d := fplja.data
    return d[n:len(d)-n]
}
```

paulangton replied:
added error handling scenarios in necessary tests :+1:

paulangton commented on prow/spyglass/spyglass.go:
```diff
@@ -0,0 +1,234 @@
+	for _, name := range haveArtifacts {
+		for i := 0; i < len(needed); i++ {
+			if name == needed[i] {
+				needed = append(needed[:i], needed[i+1:]...)
```
Agreed, would make things a lot simpler

paulangton commented on prow/spyglass/spyglass.go:
```diff
@@ -0,0 +1,234 @@
+			foundArtifactNames = append(foundArtifactNames, artifact.JobPath())
+		}
+		artElapsed := time.Since(artStart)
+		logrus.Info("Retrieved artifacts in ", artElapsed)
```
addressed above

paulangton commented on prow/spyglass/viewers/viewers.go:
```diff
@@ -0,0 +1,193 @@
+}
+
+// UnregisterViewer unregisters viewers
+func UnregisterViewer(viewerName string) {
```
That was what I had in mind when writing the method. As this is the core library I wanted it to at least have the building blocks for supporting things like external viewers in the future. They'll be thread-safe in latest.

cjwagner commented on prow/spyglass/gcsartifact_fetcher.go:
```diff
@@ -0,0 +1,261 @@
+		artifacts = append(artifacts, aName)
+	}
+	artElapsed := time.Since(artStart)
+	logrus.Infof("Listed %d GCS artifacts in %s", len(artifacts), artElapsed)
```
Thanks, I think we should keep the durations in the logs after metrics are added though. They can still be useful there when debugging.

stevekuznetsov commented on prow/spyglass/gcsartifact.go:
```diff
@@ -0,0 +1,204 @@
+}
+
+// UseContext sets the context to be used in GCS read operations
+func (a *GCSArtifact) UseContext(ctx context.Context) error {
```
Are artifacts local to a request? This seems odd, should be set at creation time

paulangton replied:
It is already set at creation time. This is for use by viewers who may want a cancellation context

cjwagner replied:
We could just always provide a cancelation context in case a viewer wants to cancel.

paulangton replied:
The point there being that there are context-specific features the viewer might want, and we dont want to reimplement context functionality on every Artifact.

stevekuznetsov commented on prow/spyglass/gcsartifact_fetcher.go:
```diff
@@ -0,0 +1,266 @@
+		}
+		resp, err := c.Get(req.String())
+		if err != nil {
+			logrus.WithField("request", req.String()).WithFields(fieldsForJob(src)).WithError(err).Error("Error in GCS XML API GET request")
```
Once you create the request, you can create the logger once so you can dedupe these constructions

paulangton replied:
xml call removed in latest

stevekuznetsov commented on prow/spyglass/gcsartifact_fetcher.go:
```diff
@@ -0,0 +1,266 @@
+	}
+	maxResults := 1000
+
+	gcsSrc := newGCSJobSource(fmt.Sprintf("gs://%s", src.jobPath()))
```
Is any GCS Job Source not going to start with `gs://`? If not, hide that in your impl and accept the job path only

paulangton replied:
the GCS parsing library provided by testgrid wants a gs:// prefix, keeping this in for consistency

stevekuznetsov replied:
I mean, add the `gs://` prefix addition _inside_ of the `newGCSJobSource()` function, don't make every caller add it.

stevekuznetsov commented on prow/spyglass/gcsartifact_fetcher_test.go:
```diff
@@ -0,0 +1,103 @@
+		expectedArtifacts []string
+	}{
+		{
+			name:         "Fetch Example CI Run #403 Artifacts",
```
Test names should be useful for people investigating the test failure -- if this test fails, what does that mean? What functionality is testing? If this is a simple happy-case, write that.

paulangton replied:
improved test case naming in latest

stevekuznetsov commented on prow/spyglass/gcsartifact_fetcher_test.go:
```diff
@@ -0,0 +1,103 @@
+		},
+	}
+
+	for _, tc := range testCases {
```
Do we want to test negative cases?

stevekuznetsov commented on prow/spyglass/gcsartifact_fetcher_test.go:
```diff
@@ -0,0 +1,103 @@
+
+// Tests listing objects associated with the current job in GCS
+// TODO: fake-gcs-server does not support the GCS XML API, Issue #8923
+func testGCSListArtifacts(t *testing.T) {
```
I thought tests needed to be exported?

paulangton replied:
exported in latest, it was intentionally unexported because of the TODO

stevekuznetsov commented on prow/spyglass/gcsartifact_test.go:
```diff
@@ -0,0 +1,126 @@
+// Tests reading at most n bytes of data from files in GCS
+func TestGCSReadAtMost(t *testing.T) {
+	buildLogArtifact := NewGCSArtifact(context.Background(), fakeGCSBucket.Object("logs/example-ci-run/403/build-log.txt"), "", "build-log.txt", 500e6)
+	testCases := []struct {
```
Corollary: add negative tests, etc... this shouldn't be only one

paulangton replied:
added negative tests

stevekuznetsov commented on prow/spyglass/gcsartifact_test.go:
```diff
@@ -0,0 +1,126 @@
+}
+
+// Tests reading all data from files in GCS
+func TestGCSReadAll(t *testing.T) {
```
Right now these tests are at the highest level, where you need:
 - connectivity to GCS (or your fake)
 - no issues reading data
 - correct logic to read all / read at most / read N/ whatever

Maybe it would help to refactor a tiny bit and allow you to unit-test that your logic works even with a fake that is much less complicated than the GCS fake you are using. Instead of using a fake GCS server, use dependency injection to add a closure that returns a specific `[]byte` on the call you're expecting, and error otherwise, etc. Reduce the scope of moving parts of these tests. The top-level test as you have it is useful, still, but harder to reason about

paulangton replied:
added a working dependency injection for `*storage.ObjectHandle` in latest

paulangton commented on prow/spyglass/gcsartifact_fetcher_test.go:
```diff
@@ -0,0 +1,103 @@
+	}
+
+	for _, tc := range testCases {
+		actualArtifacts := testAf.artifacts(tc.gcsJobSource)
```
No testAf is an actual GCSArtifactFetcher. It uses `fake-gcs-server`, but tests the real implementation.

fejta replied:
Example of something that doesn't yet have a comment

paulangton replied:
Changed in latest

paulangton commented on prow/spyglass/gcsartifact_fetcher.go:
```diff
@@ -0,0 +1,266 @@
+	}
+	err := xml.Unmarshal(content, &extracted)
+	if err != nil {
+		logrus.WithFields(fieldsForJob(src)).WithError(err).Error("Error unmarshaling artifact names from XML")
```
xml call removed in latest

paulangton commented on prow/spyglass/gcsartifact_test.go:
```diff
@@ -0,0 +1,258 @@
+			Size:            49,
+		},
+	}, "", "build-log.txt", 500e6)
+	testCases := []struct {
```
I like it, incorporating the pattern in latest

fejta commented on prow/spyglass/gcsartifact.go:
```diff
@@ -0,0 +1,240 @@
+	ctx context.Context
+}
+
+type artifactReader interface {
```
Isn't this just https://golang.org/pkg/io/#ReadCloser?

paulangton replied:
Yes but someday we may use other methods from `*storage.Reader` which is what this was meant to embed

fejta replied:
Sure,

In general I would say to make that change at that time.

We're not burning this code onto a CD. We can refactor everything in minutes with a PR. Best to wait for complexity to be required before requiring reviewers to understand and maintain that complexity.

Code that doesn't exist doesn't fail, and you're generally not going to get comments about code that hasn't been written

paulangton replied:
I definitely hold a differing opinion on that. We arent burning this code into a CD, but we also aren't working under a microscope. I think a lot of headaching and refactoring can be saved down the road by adding a little more flexibility/complexity now, especially when we can reasonably predict a few meters of road in front of us.

fejta commented on prow/spyglass/gcsartifact_test.go:
```diff
@@ -0,0 +1,248 @@
+		toRead = lenContents - offset
+		err = io.EOF
+	}
+	return &ByteReadCloser{bytes.NewReader(h.contents[offset : offset+toRead])}, err
```
Is this how ObjectHandle works? It will return a artifactReader + io.EOF if the requested length is longer?

paulangton replied:
no it will return an HTTP error with the string "EOF" contained in it. I chose to return io.EOF because this does not make a network call

fejta replied:
It sounds like a problem if your range reader differs from the actual implementation in a meaningful way?

fejta replied:
Same here?

fejta commented on prow/spyglass/gcsartifact_test.go:
```diff
@@ -0,0 +1,248 @@
+	}{
+		{
+			name:   "Test size simple",
+			handle: startedHandle,
```
Looks like startedHandle is only used in one location, so please move it into here

fejta commented on prow/spyglass/gcsartifact_test.go:
```diff
@@ -0,0 +1,248 @@
+		},
+	}
+	nonexistentHandle := &gcsArtifactHandle{fakeGCSBucket.Object("logs/example-ci-run/404/started.json")}
+	testCases := []struct {
```
This still seems to have the wrong abstraction.

startedHandle and nonexistentHandle are each only used once, suggesting they ought to go inside the handle declaration.

On the other hand you're repeating the really long contents json in both expected and startedHandle.contents, which seems error prone. Seems like this ought to be pulled out.

paulangton replied:
Removed the super long string, moved them inside the handle declaration

paulangton commented on prow/spyglass/podlogartifact_test.go:
```diff
@@ -0,0 +1,306 @@
+			if err != tc.expectedErr {
+				t.Fatalf("failed reading bytes of log. err: %v, expected err: %v", err, tc.expectedErr)
+			}
+			res = bytes.Trim(res, "\x00")
```
it doesnt anymore :)

krzyzacy commented:
@fejta @stevekuznetsov is there still any outstanding comments left? I'd prefer we get this merged and address further feedback on followups.

fejta commented on prow/spyglass/gcsartifact_fetcher.go:
```diff
@@ -0,0 +1,178 @@
+}
+
+func (h *gcsArtifactHandle) NewRangeReader(ctx context.Context, offset, length int64) (artifactReader, error) {
+	return h.ObjectHandle.NewRangeReader(ctx, offset, length)
```
Sounds like this one will return a different error than what we're unit testing? Why isn't that a problem?

paulangton replied:
addressed above, it will return the same

fejta commented:
@krzyzacy @paulangton given the large size of this PR it isn't obvious to me whether various comments have been addressed -- in particular comments from not Paul that do not have a Done, Ack, etc reply from Paul.

paulangton commented:
I had been using :+1: as a way to say that a review has been addressed and implemented (if it was a big review I would comment directly with my response) because I wasnt sure what the standard was for addressing reviews, I'll go back and add comments.

fejta commented:
Sure that works too! Just something to help me figure out what is/isn't address

fejta commented:
Ideally I use https://k8s-gubernator.appspot.com/pr/fejta to see if I need to do something, but often times people will reply and/or push a commit (to trigger testing) without actually being finished

paulangton commented on prow/spyglass/viewers/viewers.go:
```diff
@@ -0,0 +1,185 @@
+		bytesRead = bytes.Trim(bytesRead, "\x00")
+		linesInContents += int64(bytes.Count(bytesRead, []byte("\n")))
+		contents = append(bytesRead, contents...)
+		chunks += 1
```
I'm going to do this, but I'd like it to be known that I disagree with the usage of `++` over `+= 1`.

paulangton commented on prow/spyglass/gcsartifact_fetcher.go:
```diff
@@ -0,0 +1,261 @@
+	namesChan := make(chan string, maxResults*len(bodies))
+	for _, body := range bodies {
+		wg.Add(1)
+		go names(body, namesChan, src, &wg)
```
removed in final version

paulangton commented on prow/spyglass/gcsartifact_fetcher.go:
```diff
@@ -0,0 +1,266 @@
+		artifacts = append(artifacts, aName)
+	}
+	artElapsed := time.Since(artStart)
+	logrus.WithField("duration", artElapsed).Infof("Listed %d GCS artifacts.", len(artifacts))
```
going to leave this out for now, if I add more metrics (for future PRs) i will add this

paulangton commented on prow/spyglass/gcsartifact_test.go:
```diff
@@ -0,0 +1,126 @@
+			expected: []byte("Oh "),
+		},
+	}
+	for _, tc := range testCases {
```
better test coverage in latest

paulangton commented:
Reviewers: (@stevekuznetsov @fejta @krzyzacy @BenTheElder @cjwagner) all reviews have been addressed with a thumbs up (meaning it was implemented) or a comment (implemented or in discussion).

stevekuznetsov commented on prow/spyglass/spyglass.go:
```diff
@@ -0,0 +1,239 @@
+)
+
+var (
+	// These are artifacts that we guarantee will be created with every job.
```
Clone log becomes build log if we fail in `clonerefs`

paulangton commented on prow/spyglass/spyglass_test.go:
```diff
@@ -0,0 +1,186 @@
+	defer fakeGCSServer.Stop()
+	fakeGCSClient := fakeGCSServer.Client()
+	fakeGCSBucket = fakeGCSClient.Bucket("test-bucket")
+	testAf = NewGCSArtifactFetcher(fakeGCSClient, fakeGCSServer.URL()+"/storage/v1", false)
```
At some point there is going to have to be a global, unless we want to spin up a new server for each test function. Im going to leave the server global and make the client and artifact fetchers local

stevekuznetsov replied:
What is the demonstrated issue with test-local servers?

stevekuznetsov replied:
Also, if the server is heavyweight, dependency injection that allows us to hide the server connection behind an interface would be helpful and allow us to start the server in an integration test only.

stevekuznetsov commented on prow/spyglass/gcsartifact_fetcher.go:
```diff
@@ -0,0 +1,178 @@
+// GCSArtifactFetcher contains information used for fetching artifacts from GCS
+type GCSArtifactFetcher struct {
+	client      *storage.Client
+	xmlEndpoint string
```
Unused?

stevekuznetsov commented on prow/spyglass/gcsartifact_fetcher_test.go:
```diff
@@ -0,0 +1,168 @@
+}
+
+// Tests listing objects associated with the current job in GCS
+func TestArtifacts_ListGCS(t *testing.T) {
```
@fejta made the comment as well, but whenever possible I want to be able to scope the amount my brain needs to think to everything between the `{}` on this test function only. The more I need to read outside, the harder it is for me to consider what the implications are and the harder it is to enforce some invariant. Instead of enforcing "we list artifacts correctly" we are now enforcing "given the correct global setup for our fake, given that testmain still sets the global, given that .... " and that makes it easier to break this test without actually changing the functionality it is supposed to be testing. Can you fabricate simple `jobSource`s in this test that you construct to return a simple list, etc?

stevekuznetsov commented on prow/spyglass/gcsartifact_test.go:
```diff
@@ -0,0 +1,248 @@
+)
+
+type ByteReadCloser struct {
+	io.Reader
```
It's going to do dynamic dispatch anyway at runtime so it doesn't matter what is slotted in there

stevekuznetsov commented on prow/spyglass/podlogartifact_test.go:
```diff
@@ -0,0 +1,301 @@
+	"k8s.io/test-infra/prow/spyglass/viewers"
+)
+
+type fakePodLogJAgent struct {
```
?

paulangton replied:
added a comment, hopefully that will help :slightly_smiling_face:

stevekuznetsov replied:
No, hard to tell what _you_ mean :P

stevekuznetsov commented on prow/spyglass/spyglass_test.go:
```diff
@@ -0,0 +1,186 @@
+	for i := 0; i < 300; i++ {
+		longLog += "here a log\nthere a log\neverywhere a log log\n"
+	}
+	fakeGCSServer := fakestorage.NewServer([]fakestorage.Object{
```
Each test uses at most one of these little bits. Make them data-driven from the test. Let the test call a constructor and pass in string literals or whatever, where necessary. Drive the functionality of your fake from data in your test matrix. Don't see "duplicated" test fakes as an opportunity to DRY things out, as that starts creating weird dependencies between tests and starts to broaden the scope of your tests from "X functionality works" to "X functionality works and the shared test fake is set-up in the correct way". I should not be able to break test Y unless I break functionality Y (generic fake harnesses aside). See how the fake clients for k8s let you seed them with behavior at creation time.

paulangton commented on prow/spyglass/spyglass.go:
```diff
@@ -0,0 +1,239 @@
+}
+
+// Views gets all views of all artifact files matching each regexp with a registered viewer
+func (s *Spyglass) Views(matchCache map[string][]string) []Lens {
```
Yes, added in latest

paulangton commented on prow/spyglass/spyglass.go:
```diff
@@ -0,0 +1,239 @@
+}
+
+// Refresh reloads the html view for a given view and a set of objects
+func (s *Spyglass) Refresh(src string, podName string, sizeLimit int64, viewReq *ViewRequest) (Lens, error) {
```
Im going to push testing this to another PR as well

krzyzacy commented:
Let's get this in, and we can address further feedback on tests, aka https://github.com/kubernetes/test-infra/issues/9020 (I'd say spyglass has pretty good test coverages, and give we'll iterate on the logic base on the actual behavior, we'll probably need to tune tests a lot as well)

Also I'm pretty sure we need to iterate on https://github.com/kubernetes/test-infra/pull/8922 before it can be fully deployed.

/lgtm
/hold

(gonna hold cancel by EoD)

krzyzacy commented:
and takes 10min to bypass the :unicorn: 

/hold cancel
